src/core/metrics.cc

```diff
@@ -163,205 +182,307 @@ Metrics::EnableGPUMetrics()
-    nvml_thread_exit_.store(false);
-    nvml_thread_.reset(new std::thread([this, dcnt] {
-      // Stop attempting any metric the fails multiple consecutive
+  // Periodically send the DCGM metrics...
+  if (available_gpu_ids.size() > 0) {
+    dcgmHandle_t handle = dcgm_handle_;
+    dcgmGpuGrp_t groupId = groupId_;
+    dcgm_thread_exit_.store(false);
+    dcgm_thread_.reset(new std::thread([this, available_gpu_ids, handle,
+                                        groupId] {
+      int available_gpu_count = available_gpu_ids.size();
+      // Stop attempting metrics if they fail multiple consecutive
       // times for a device.
       constexpr int fail_threshold = 3;
-      std::vector<int> power_limit_fail_cnt(dcnt);
-      std::vector<int> power_usage_fail_cnt(dcnt);
-      std::vector<int> energy_fail_cnt(dcnt);
-      std::vector<int> util_fail_cnt(dcnt);
-      std::vector<int> mem_fail_cnt(dcnt);
-
-      unsigned long long last_energy[dcnt];
-      for (int didx = 0; didx < dcnt; ++didx) {
+      std::vector<int> power_limit_fail_cnt(available_gpu_count);
+      std::vector<int> power_usage_fail_cnt(available_gpu_count);
+      std::vector<int> energy_fail_cnt(available_gpu_count);
+      std::vector<int> util_fail_cnt(available_gpu_count);
+      std::vector<int> mem_fail_cnt(available_gpu_count);
+      std::vector<int> cuda_available_cnt(available_gpu_count);
+
+      unsigned long long last_energy[available_gpu_count];
+      for (int didx = 0; didx < available_gpu_count; ++didx) {
         last_energy[didx] = 0;
       }
-
-      while (!nvml_thread_exit_.load()) {
-        std::this_thread::sleep_for(std::chrono::milliseconds(2000));
-
-        for (int didx = 0; didx < dcnt; ++didx) {
-          nvmlDevice_t gpu = nvml_device_[didx];
-
-          // Power limit
-          if (power_limit_fail_cnt[didx] < fail_threshold) {
-            unsigned int power_limit;
-            nvmlReturn_t nvmlerr =
-                nvmlDeviceGetPowerManagementLimit(gpu, &power_limit);
-            if (nvmlerr == NVML_SUCCESS) {
-              power_limit_fail_cnt[didx] = 0;
-            } else {
-              LOG_WARNING << "failed to get power limit for GPU " << didx
-                          << ": " << nvmlErrorString(nvmlerr);
-              power_limit = 0;
+      size_t field_count = 6;
+      unsigned short util_flag =
+          standalone_ ? DCGM_FI_PROF_GR_ENGINE_ACTIVE : DCGM_FI_DEV_GPU_UTIL;
+      unsigned short fields[field_count] = {
+          DCGM_FI_DEV_POWER_MGMT_LIMIT,          // power limit, watts
+          DCGM_FI_DEV_POWER_USAGE,               // power usage, watts
+          DCGM_FI_DEV_TOTAL_ENERGY_CONSUMPTION,  // Total energy consumption, mJ
+          util_flag,                             // util ratio, 1 = 1%
+          DCGM_FI_DEV_FB_USED,                   // Frame buffer used, MB
+          DCGM_FI_DEV_FB_TOTAL,                  // Frame buffer used, MB
+      };
+
+      char fieldName[16] = "myfields";
```

Comment:
Why is this 16 again?

Reply:
Also use "field_group"
In general keep possession words out of variable names unless they signify meaning,